### PR TITLE
Collect response headers

### PR DESCRIPTION
**Pull request by shrey-garg**

### CODE REVIEW

**Comment by shrey-garg** [comment]
after #111

File: service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java [hunk 1]
```diff
@@ -164,4 +178,21 @@ public static void shutDown() {
         isDebug.remove();
         debugResponses.remove();
     }
+
+    static Map<String, Queue<String>> getCollectedHeaders() {
+        return collectedHeaders.get();
+    }
+
+    public List<String> getCollectedHeaders(String header) {
```
Comment (ishwar-kumar) [inline service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java]: static

File: service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java [hunk 2]
```diff
@@ -46,40 +44,56 @@ protected Boolean initialValue() {
...
-    public static void initialize() {
+    public static void initialize(List<String> responseHeadersToCollect) {
         if (isImmutable.get()) {
             throw new UnsupportedOperationException();
         }
 
         context.remove();
         isImmutable.set(false);
+
+        if (responseHeadersToCollect.isEmpty()) {
```
Comment (pmohankumar) [inline service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java]: Good to have a null check here as this comes from config provided by app

File: service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java [hunk 3]
```diff
@@ -46,40 +44,56 @@ protected Boolean initialValue() {
...
+            return;
+        }
+
+        collectedHeaders = ThreadLocal.withInitial(() -> {
```
Comment (pmohankumar) [inline service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java]: We can still make this final and initialize in static declaration. But here, we can just get map from thread local and add to it (as the map itself isn't immutable)

File: service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java [hunk 4]
```diff
@@ -164,4 +178,21 @@ public static void shutDown() {
         isDebug.remove();
         debugResponses.remove();
     }
+
+    static Map<String, Queue<String>> getCollectedHeaders() {
+        return collectedHeaders.get();
+    }
+
+    public static List<String> getCollectedHeaders(String header) {
+        return new ArrayList<>(getCollectedHeaders().get(header.toLowerCase()));
```
Comment (pmohankumar) [inline service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java]: Use optional as map might not have header which app is passing here. Passing null to arraylist will result in NPE

**Pull request merged**